### PR TITLE
Added initial changes to support multiple Merkle trees in the storage.

### PR DESCRIPTION
**Pull request by HristoStaykov**
Added assertion to check for collisions + fix.
Added the address field to the InternalNodeKey's serialization.

* **Problem Overview**  
 Initial work to support multiple Merkle trees in storage. This will be needed when we build a MPT per account and a global MPT where the keys will be the account addresses and the values will be the account storage Merkle roots.
* **Testing Done**  
  All unit tests pass + added a non empty address in unit test
  serialization_unit_test/stale_db_key_internal for better
  code coverage. 
Ran locally 200 times the property based test **sparse_merkle_storage_db_adapter_property_test**

### CODE REVIEW

File: kvbc/cmf/categorized_kvbc_msgs.cmf [hunk 1]
```diff
@@ -201,6 +201,7 @@ Msg BatchedInternalNodeKey 5002 {
     # The version of this key is the *tree* version, not block version
     uint64 version
     NibblePath path
+    string address
```
Comment (vrancurel) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:204]: What is the address ? and why do we need it in the internal nodes ?

Comment (vrancurel) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:204]: What exactly is the function of this message ?

Comment (vrancurel) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:204]: Why do we need to store the (account?) address in the node key ?

Comment (vrancurel) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:204]: OK so we serialize all Merkle tree in the same collection, and we distinguish the merkle tree node keys by the account address ?

Comment (HristoStaykov) [reply kvbc/cmf/categorized_kvbc_msgs.cmf:204]: Yes, we can have all MPT-s (the Global state, which will have an empty address field and each account's tree) in the same storage space (CF).

File: kvbc/include/merkle_tree_serialization.h [hunk 2]
```diff
@@ -107,6 +113,7 @@ inline std::string serializeImp(const sparse_merkle::Hash &hash) {
 inline void serializeImp(const sparse_merkle::InternalNodeKey &key, std::string &out) {
   serializeImp(key.version().value(), out);
   serializeImp(key.path(), out);
+  serializeImp(key.address(), out);
```
[the remark this answers is not in the document]
Comment (HristoStaykov) [reply kvbc/include/merkle_tree_serialization.h:116]: Yes, this change makes sense for a better storage organization. I will move the address to the top.

**Review by teoparvanov — APPROVED** [review]
Great job, thanks @HristoStaykov !
(please include the additional property-based testing you did over the weekend in the "Testing done" section)

File: kvbc/include/merkle_tree_serialization.h [hunk 3]
```diff
@@ -105,6 +111,7 @@ inline std::string serializeImp(const sparse_merkle::Hash &hash) {
 }
 
 inline void serializeImp(const sparse_merkle::InternalNodeKey &key, std::string &out) {
+  serializeImp(key.address(), out);
```
Comment (vrancurel) [inline kvbc/include/merkle_tree_serialization.h:114]: Is it possible to have a prefix to distinguish for other types of nodes ? (we would have all the Merkle keys in the same CF).

**Review by vrancurel — CHANGES_REQUESTED** [review]

File: kvbc/include/categorization/block_merkle_category.h [hunk 4]
```diff
@@ -202,7 +202,7 @@ class BlockMerkleCategory {
     Reader(const storage::rocksdb::NativeClient& db) : db_{db} {}
 
     // Return the latest root node in the system.
-    sparse_merkle::BatchedInternalNode get_latest_root() const override;
+    sparse_merkle::BatchedInternalNode get_latest_root(std::string address = "") const override;
```
Comment (vrancurel) [inline kvbc/include/categorization/block_merkle_category.h:205]: Same here, etc.

**Review by vrancurel — CHANGES_REQUESTED** [review]

File: kvbc/cmf/categorized_kvbc_msgs.cmf [hunk 5]
```diff
@@ -199,6 +199,7 @@ Msg NibblePath 5001 {
 
 Msg BatchedInternalNodeKey 5002 {
     # The version of this key is the *tree* version, not block version
+    string address
```
[the remark this answers is not in the document]
Comment (HristoStaykov) [reply kvbc/cmf/categorized_kvbc_msgs.cmf]: Replaced globally.

File: kvbc/test/sparse_merkle/tree_test.cpp [hunk 6]
```diff
@@ -274,7 +274,7 @@ TEST(tree_tests, insert_single_leaf_node) {
   // Only a single node was created. It's key is made up of an empty nibble
   // path, since there are no collisions requring going deeper into the tree.
   auto& root = batch.internal_nodes.at(0);
-  ASSERT_EQ(InternalNodeKey(1, NibblePath()), root.first);
+  ASSERT_EQ(InternalNodeKey("", 1, NibblePath()), root.first);
```
Comment (cloudnoize) [inline kvbc/test/sparse_merkle/tree_test.cpp:277]: do you have tests that create multiple trees?

Comment (HristoStaykov) [reply kvbc/test/sparse_merkle/tree_test.cpp:277]: yes, a test covering multiple MPT-s is implemented here [PR-3007](https://github.com/vmware/concord-bft/pull/3007)

**Pull request merged**